### ISSUE TITLE
Invalid oom_adj value on the worker processes

### ISSUE THREAD

**Issue by green-mdoerflinger** [phusion/passenger#1631]
Good day,

some years ago we reported an error in the way passenger sets the oom_adj value on the worker processes (this was fixed in #843). We have recently upgraded to Passenger 5 on our production systems (using 5.0.16 at the moment) and it seems the error is back.

I.e. all worker processes inherit the oom_adj of -17 from the watchdog which excludes them from the oom killer. As the rails processes can become excessively large this is a very big problem for us (system will kill mysqld and other very important processes).

Regards,
Marc

**Comment by FooBarWidget** [comment]
Hi,

I can't seem to reproduce this issue. The Watchdog sets its oom_adj to 17, but the Core (a subprocess of Watchdog which will further spawn the Rails apps) has a proper oom_adj value of 0.

**Comment by green-mdoerflinger** [comment]
The issue are the viewer processes (Passenger RackApp), not the core. Those processes have a oom_adj of -17.

**Comment by FooBarWidget** [comment]
Yeah, but the Passenger RackApp should inherit its oom_adj from the Core. Can you check what oom_adj value the Core process has?

**Comment by green-mdoerflinger** [comment]
I just checked using version passenber 5.0.21 and the core as well as the viewers are having oom_adj -17. We are running on Debian 8.

**Comment by green-mdoerflinger** [comment]
Sorry for the close, clicked the wrong button. ;-)

**Comment by FooBarWidget** [comment]
That tells me that the oom_adj adjustment isn't working. Very odd. I'll need to try to reproduce this on Debian 8.

Did you install Passenger from gem/source or are you using our Debian packages?

**Comment by green-mdoerflinger** [comment]
I installed it from gem using

> gem install passenger -v 5.0.21 -V --no-ri --no-rdoc

(using gem 2.2.2 and ruby 2.1.5p273)

**Comment by green-mdoerflinger** [comment]
Any idea when this gets fixed? We would really like to upgrade to passenger 5, but without that fix its a no-go for us. Thanks.

**Comment by FooBarWidget** [comment]
I'll have another look for 5.0.27. No luck reproducing the issue so far.

**Comment by OnixGH** [comment]
@green-mdoerflinger are you sure you aren't launching Passenger from a process that has `oom_adj -17` itself?

Because in that case it is expected behavior for all the subprocesses to inherit that same -17 score.

If I start passenger (root) in a Debian 8 shell with oom_adj 0, then everything looks good: the watchdog sets itself to -17 and the rest of the processes inherit the 0 from the shell.

**Comment by green-mdoerflinger** [comment]
I think i have left out a little detail that might help reproducing the issue: we are running the apache2 passenger module not standalone. The apache2 is the default version that comes with Debian 8 (started using `service apache2 start`). Module configuration looks like this

```
LoadModule passenger_module /var/lib/gems/2.1.0/gems/passenger-5.0.26/buildout/apache2/mod_passenger.so
PassengerRoot /var/lib/gems/2.1.0/gems/passenger-5.0.26
PassengerDefaultRuby /usr/bin/ruby2.1
```

I just tried again with passenger 5.0.26 and still oom_adj is -17.

**Comment by OnixGH** [comment]
@green-mdoerflinger Did you see my question? What I'm trying to say is that if you launch Passenger from a context that has `oom_adj -17`, then it is normal behavior that Passenger and its workers will inherit that same score.

For example, I can reproduce everything having -17 if I set my shell to -17 and then run apache2:
```
> ps
25315 pts/0    00:00:00 bash

> echo -17 > /proc/25315/oom_adj
> apache2ctl -k start
> passenger-memory-stats
25952  421.0 MB  1.1 MB   Passenger watchdog
25959  630.6 MB  1.6 MB   Passenger core
25964  429.3 MB  1.2 MB   Passenger ust-router
26106  80.7 MB   5.3 MB   Passenger AppPreloader: /home/onix/simple
26127  212.6 MB  5.3 MB   Passenger RubyApp: /home/onix/simple (development)

> cat /proc/25952/oom_adj ; cat /proc/25959/oom_adj ; cat /proc/26127/oom_adj 
-17
-17
-17
```

But that's expected. If I set my shell to 0 then only the watchdog is -17 (also as expected):
```
> echo 0 > /proc/25315/oom_adj
> apache2ctl -k start
> passenger-memory-stats
26266  421.0 MB  1.1 MB   Passenger watchdog
26269  630.6 MB  1.6 MB   Passenger core
26275  429.3 MB  1.2 MB   Passenger ust-router
26409  80.7 MB   5.3 MB   Passenger AppPreloader: /home/onix/simple
26430  212.6 MB  5.4 MB   Passenger RubyApp: /home/onix/simple (development)

>  cat /proc/26266/oom_adj ; cat /proc/26269/oom_adj ; cat /proc/26430/oom_adj 
-17
0
0
```

In conclusion, you should check the oom_adj score of the parent context from which you're launching. So far Passenger 5/Apache2 is behaving as expected on a fresh Debian 8.

**Comment by green-mdoerflinger** [comment]
I tried the same thing, except i used `service apache2 start` to start my apache, but unfortunately i still get oom_adj -17 on the viewers. See below for the log, you can see that my bash process (1821) has an oom_adj of 0. 

```
root@fast:~# ps
  PID TTY          TIME CMD
 1819 pts/0    00:00:00 sudo
 1820 pts/0    00:00:00 su
 1821 pts/0    00:00:00 bash
18222 pts/0    00:00:00 ps
root@fast:~# cat /proc/1821/oom_adj 
0
root@fast:~# passenger-memory-stats 
Version: 5.0.26
Date   : 2016-03-22 10:34:34 +0100

---------- Apache processes ----------
PID    PPID   VMSize    Private  Name
--------------------------------------
17878  1      116.8 MB  0.7 MB   /usr/sbin/apache2 -k start
17911  17878  591.3 MB  3.0 MB   /usr/sbin/apache2 -k start
17912  17878  463.4 MB  3.0 MB   /usr/sbin/apache2 -k start
### Processes: 3
### Total private dirty RSS: 6.65 MB


-------- Nginx processes --------

### Processes: 0
### Total private dirty RSS: 0.00 MB


----- Passenger processes ------
PID    VMSize     Private  Name
--------------------------------
17881  421.1 MB   1.1 MB   Passenger watchdog
17884  1032.6 MB  2.2 MB   Passenger core
17889  429.5 MB   1.2 MB   Passenger ust-router
17983  284.6 MB   16.9 MB  Passenger AppPreloader: /fast/fast_9203/viewer
18004  352.3 MB   55.6 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
18012  352.4 MB   48.4 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
18019  352.5 MB   17.0 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
### Processes: 7
### Total private dirty RSS: 142.41 MB
root@fast:~# cat /proc/17881/oom_adj; cat /proc/17884/oom_adj; cat /proc/18004/oom_adj 
-17
-17
-17

```

**Comment by OnixGH** [comment]
And what happens if you use apache2ctl?

**Comment by green-mdoerflinger** [comment]
The same 

```
root@fast:~# apache2ctl -k start
\AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 127.0.1.1. Set the 'ServerName' directive globally to suppress this message
root@fast:~# passenger-memory-stats 
Version: 5.0.26
Date   : 2016-03-22 10:45:31 +0100

---------- Apache processes -----------
PID    PPID   VMSize     Private  Name
---------------------------------------
18522  1      116.8 MB   0.7 MB   /usr/sbin/apache2 -k start
18553  18522  1039.4 MB  3.1 MB   /usr/sbin/apache2 -k start
18554  18522  591.3 MB   3.0 MB   /usr/sbin/apache2 -k start
### Processes: 3
### Total private dirty RSS: 6.79 MB


-------- Nginx processes --------

### Processes: 0
### Total private dirty RSS: 0.00 MB


----- Passenger processes ------
PID    VMSize     Private  Name
--------------------------------
18523  421.1 MB   1.1 MB   Passenger watchdog
18530  1032.6 MB  2.2 MB   Passenger core
18537  429.5 MB   1.2 MB   Passenger ust-router
18610  284.7 MB   17.2 MB  Passenger AppPreloader: /fast/fast_9203/viewer
18628  352.4 MB   56.3 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
18635  352.5 MB   36.0 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
18642  352.6 MB   17.4 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
### Processes: 7
### Total private dirty RSS: 131.47 MB
root@fast:~# cat /proc/18523/oom_adj; cat /proc/18530/oom_adj; cat /proc/18628/oom_adj 
-17
-17
-17
```

**Comment by OnixGH** [comment]
Does your Apache processes also have -17?

If that is so there's nothing much Passenger can do about it, it just inherits, and you should investigate why apache is started with -17 on your system (I still don't get this on a fresh Debian 8, not even with `service apache2 start`).

**Comment by green-mdoerflinger** [comment]
No, the apache processes all have oom_adj 0 too

```
root@fast:~# passenger-memory-stats 
Version: 5.0.26
Date   : 2016-03-22 11:04:32 +0100

---------- Apache processes -----------
PID    PPID   VMSize     Private  Name
---------------------------------------
18522  1      116.8 MB   0.7 MB   /usr/sbin/apache2 -k start
18553  18522  1039.4 MB  3.1 MB   /usr/sbin/apache2 -k start
18554  18522  591.3 MB   3.0 MB   /usr/sbin/apache2 -k start
### Processes: 3
### Total private dirty RSS: 6.80 MB


-------- Nginx processes --------

### Processes: 0
### Total private dirty RSS: 0.00 MB


----- Passenger processes ------
PID    VMSize     Private  Name
--------------------------------
18523  421.1 MB   1.1 MB   Passenger watchdog
18530  1032.6 MB  2.2 MB   Passenger core
18537  429.5 MB   1.2 MB   Passenger ust-router
18610  284.7 MB   17.2 MB  Passenger AppPreloader: /fast/fast_9203/viewer
18628  352.4 MB   56.3 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
18635  352.5 MB   36.0 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
18642  352.6 MB   17.4 MB  Passenger RubyApp: /fast/fast_9203/viewer (production_test)
### Processes: 7
### Total private dirty RSS: 131.48 MB
root@fast:~# cat /proc/18522/oom_adj; cat /proc/18553/oom_adj;cat /proc/18554/oom_adj
0
0
0
```

**Comment by green-mdoerflinger** [comment]
Note: i get the behaviour on multiple different systems, all running plain Debian 8.

**Comment by OnixGH** [comment]
Aha, ok, if the apache process have 0 then that does strongly point to something in the watchdog.

Is there some way you can supply a reproducible case, like a VM or a digitalocean droplet? I tried with digitalocean / Debian 8.3 x64 and it doesn't happen for me :(.

**Comment by green-mdoerflinger** [comment]
I can provide a virtualbox image, but i need to prepare it first (remove part of our commercial software from it). Would that work for you?

**Comment by OnixGH** [comment]
Yep, VirtualBox would be excellent, thanks!

**Comment by green-mdoerflinger** [comment]
I have the vm ready to download from my google drive. 
Could you please provide me with an email address or another way of sharing you the link to the vm image download privately. Thanks.

**Comment by OnixGH** [comment]
You can send it to us at team@phusionpassenger.com.

**Comment by green-mdoerflinger** [comment]
I just sent an email with the link and details about the vm.

**Comment by OnixGH** [comment]
We have received / downloaded it, thanks!

**Comment by OnixGH** [comment]
@green-mdoerflinger aha, you have `PassengerUserSwitching off` in your config; this causes the watchdog to drop privileges after setting -17 and renders the child processes unable (Permission Denied) to change their inherited score back to the pre-watchdog 0.

Workaround: if you turn user switching on, things should work as expected with the OOM score. You are [already using](https://www.phusionpassenger.com/library/deploy/apache/user_sandboxing.html) `PassengerUser` and `PassengerGroup`, with the fallback `PassengerDefaultUser` and `PassengerDefaultGroup`, so the only difference would be that the Watchdog and the Core run as root (and can adjust the oom score).

Meanwhile we've discovered that while neither watchdog nor its child can change their own OOM score after dropping privileges (user switching off), it seems the watchdog can still change the child OOM, so a rewrite from that angle could fix this.

**Comment by green-mdoerflinger** [comment]
Thanks for the analysis. I will look into that solution and check if it would work for us. 
But if you could introduce a fix that it still works with `PassengerUserSwitching off` that would be great.
